[Gtk][WPE][Skia] Use SkPicture(Recorder) instead of DisplayList(Recorder) for paint command recording in multi-threaded rendering mode

nikolaszimmermann opened the pull request:
#### fca89822979c51c7dbfe019c6808fb3e8cccbefc
<pre>
[Gtk][WPE][Skia] Use SkPicture(Recorder) instead of DisplayList(Recorder) for paint command recording in multi-threaded rendering mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=289456">https://bugs.webkit.org/show_bug.cgi?id=289456</a>

Reviewed by Carlos Garcia Campos.

For multi-threaded rendering Gtk/WPE ports serialized the painting
commands for a given layer and tile into a DisplayList, and dispatched
a task to a WorkerPool to replay the DisplayList in a secondary thread.

The cross-thread usage of DisplayLists is fragile, and the way we use
DisplayList wouldn&apos;t work on Apple ports, due to timing differences when
a new frame rendering starts. While DisplayList (and its ResourceHeap) protects
the referenced WebCore objects (such as Gradient / Pattern etc.) from
destruction, we&apos;re not protected from mutations between the recording of
a DisplayList and the replay, later on in time.

Furthermore, the Gtk/WPE ports currently depend on specific
synchronization primitives (creating fences in the accelerated rendering
case during recording, and waiting for the fences in the replay, before
attempting to use the fence-protected GPU resource) that were built only
into DisplayList. That was a bad idea, the Skia ports should be able to
properly paint in both CPU and GPU mode, independent of whether
DisplayList is used or not.

This PR moves the Gtk/WPE ports away from DisplayList recording to use a
Skia built-in functionality instead: SkPicture and SkPictureRecorder.
While this doesn&apos;t yet solve the mutation problem in a general way, it
removes a lot of potential problems, when we start to paint the next
frame earlier, then we do right now. Furthermore it resolves some
general design issues Kimmo detected with the current code -- namely
the sprinkling of USE(SKIA) throughout ImageBuffer/ImageBufferBackend
and NativeImage &amp; co.

Follow-up patches will remove the Skia-specific functionality that was
added to ImageBuffer &amp; co to make threaded DisplayList rendering
possible for our ports (albeit only under limited constraints, that we
want to get rid of in the long-term).

Covered by existing tests.

* LayoutTests/fast/backgrounds/background-clip-border-area-alpha.html:
Adjust for slight pixel test differences -- no visible change.
* LayoutTests/fast/multicol/transform-inside-opacity.html:
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/root-element-background-image-transparency-001.html:
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/root-element-background-image-transparency-002.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/clip-002.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/clip-004.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/clip-005.html:
* LayoutTests/svg/as-background-image/svg-transformed-background.html:
* LayoutTests/svg/repaint/buffered-rendering-static-image.html:
* Source/WebCore/platform/Skia.cmake:
* Source/WebCore/platform/SourcesSkia.txt:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::surface const):
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
(WebCore::ImageBufferBackend::surface const):
(WebCore::ImageBufferBackend::waitForAcceleratedRenderingFenceCompletion):
* Source/WebCore/platform/graphics/NativeImage.h:
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::drawNativeImageInternal):
(WebCore::GraphicsContextSkia::setupFillSource):
(WebCore::GraphicsContextSkia::setupStrokeSource):
(WebCore::GraphicsContextSkia::clipToImageBuffer):
(WebCore::createAcceleratedSurface):
(WebCore::GraphicsContextSkia::drawPattern):
(WebCore::GraphicsContextSkia::beginRecording):
(WebCore::GraphicsContextSkia::endRecording):
(WebCore::GraphicsContextSkia::createAcceleratedRenderingFenceIfNeeded):
(WebCore::GraphicsContextSkia::trackAcceleratedRenderingFenceIfNeeded):
(WebCore::GraphicsContextSkia::drawSkiaImage): Deleted.
(WebCore::GraphicsContextSkia::drawFilteredImageBuffer): Deleted.
(WebCore::GraphicsContextSkia::setupFillSource const): Deleted.
(WebCore::GraphicsContextSkia::setupStrokeSource const): Deleted.
(WebCore::GraphicsContextSkia::drawSkiaPattern): Deleted.
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.h:
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.h:
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaSurfaceBackend.h:
* Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.cpp:
(WebCore::SkiaPaintingEngine::postPaintingTask):
(WebCore::SkiaPaintingEngine::recordDisplayList const): Deleted.
(WebCore::SkiaPaintingEngine::paintDisplayListIntoBuffer): Deleted.
* Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.h:
* Source/WebCore/platform/graphics/skia/SkiaReplayCanvas.cpp: Added.
(WebCore::SkiaReplayCanvas::SkiaReplayCanvas):
(WebCore::SkiaReplayCanvas::create):
(WebCore::SkiaReplayCanvas::waitForRenderingCompletionAndRewrapImageIfNeeded):
(WebCore::SkiaReplayCanvas::invokeDrawFunctionWithImage):
(WebCore::SkiaReplayCanvas::invokeDrawFunctionWithPaint):
(WebCore::SkiaReplayCanvas::invokeDrawFunctionWithShader):
(WebCore::SkiaReplayCanvas::onClipShader):
(WebCore::SkiaReplayCanvas::onDrawArc):
(WebCore::SkiaReplayCanvas::onDrawAtlas2):
(WebCore::SkiaReplayCanvas::onDrawBehind):
(WebCore::SkiaReplayCanvas::onDrawDRRect):
(WebCore::SkiaReplayCanvas::onDrawEdgeAAImageSet2):
(WebCore::SkiaReplayCanvas::onDrawGlyphRunList):
(WebCore::SkiaReplayCanvas::onDrawImage2):
(WebCore::SkiaReplayCanvas::onDrawImageLattice2):
(WebCore::SkiaReplayCanvas::onDrawImageRect2):
(WebCore::SkiaReplayCanvas::onDrawOval):
(WebCore::SkiaReplayCanvas::onDrawPaint):
(WebCore::SkiaReplayCanvas::onDrawPoints):
(WebCore::SkiaReplayCanvas::onDrawPatch):
(WebCore::SkiaReplayCanvas::onDrawPath):
(WebCore::SkiaReplayCanvas::onDrawRRect):
(WebCore::SkiaReplayCanvas::onDrawRect):
(WebCore::SkiaReplayCanvas::onDrawRegion):
(WebCore::SkiaReplayCanvas::onDrawSlug):
(WebCore::SkiaReplayCanvas::onDrawTextBlob):
(WebCore::SkiaReplayCanvas::onDrawVerticesObject):
* Source/WebCore/platform/graphics/skia/SkiaReplayCanvas.h: Added.

Canonical link: <a href="https://commits.webkit.org/292639@main">https://commits.webkit.org/292639@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eab77ffc5f9861606d8945ebff4ec8c0a99ebc7e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96627 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16241 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6348 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101699 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47147 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98672 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16537 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24675 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73666 "Found 1 new test failure: fast/multicol/layers-split-across-columns.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30860 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99630 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12439 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87372 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53975 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12198 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5165 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46475 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82299 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5258 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103723 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23694 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17270 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82686 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23945 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83421 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82097 "Found 3 new API test failures: /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print-errors, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/close-after-print (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20607 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26718 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4241 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17168 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23657 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28812 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23316 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26796 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25057 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.h:
```diff
@@ -27,25 +27,36 @@
 
     const DestinationColorSpace& colorSpace() const final;
 
+    HashMap<const SkImage*, std::unique_ptr<GLFence>>&& releaseImageFences();
```
So this doesn't actually release the fences, it just gives the opportunity for the caller to steal the fences.
So you need to have the signature as 
`HashMap<const SkImage*, std::unique_ptr<GLFence>> releaseImageFences();`

The implementation is fine (return move or return exchange).

nikolaszimmermann replied:
Good catch, will adapt.

kkinnunen-apple reviewed (commented):
Looks very great!

carlosgcampos commented on Source/WebCore/platform/graphics/ImageBuffer.cpp:
```diff
@@ -356,23 +363,41 @@ RefPtr<ImageBuffer> ImageBuffer::sinkIntoBufferForDifferentThread(RefPtr<ImageBu
+    if (fence) {
+        fence->serverWait();
+        fence = nullptr;
+    }
```
I think we should wait after making the context current

nikolaszimmermann replied:
We do that -- this function only exists for the sake of sharing code between the two createAcceleratedRenderingFenceIfNeeded() variants and is not mean to be used without prior checking that the imge is texture backed.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -1010,35 +1000,114 @@ void GraphicsContextSkia::drawSkiaPattern(const sk_sp<SkImage>& image, const Int
 }
 
-void GraphicsContextSkia::drawPattern(NativeImage& nativeImage, const FloatRect& destRect, const FloatRect& tileRect, const AffineTransform& patternTransform, const FloatPoint& phase, const FloatSize& spacing, ImagePaintingOptions options)
+HashMap<const SkImage*, std::unique_ptr<GLFence>> GraphicsContextSkia::releaseImageFences()
```
release confuses me a bit, I would use takeImageFences() or stealImageFences() instead.

nikolaszimmermann replied:
I'd go for takeImageFences() -- but it's obsoleted by your beginRecording / endRecording proposal, which is even nicer.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -1010,35 +1000,114 @@ void GraphicsContextSkia::drawSkiaPattern(const sk_sp<SkImage>& image, const Int
+    if (!image)
+        return nullptr;
+
+    auto* glContext = PlatformDisplay::sharedDisplay().skiaGLContext();
```
I guess we should return early here if the render mode is not accelerated, before trying to get the gl context.

nikolaszimmermann replied:
See above.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -1010,35 +1000,114 @@ void GraphicsContextSkia::drawSkiaPattern(const sk_sp<SkImage>& image, const Int
+    if (!surface)
+        return nullptr;
+
+    auto* glContext = PlatformDisplay::sharedDisplay().skiaGLContext();
```
Same here.

nikolaszimmermann replied:
Ditto :-)

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -1010,35 +1000,114 @@ void GraphicsContextSkia::drawSkiaPattern(const sk_sp<SkImage>& image, const Int
+    auto* grContext = PlatformDisplay::sharedDisplay().skiaGrContext();
+    RELEASE_ASSERT(grContext);
+
+    grContext->flush(image);
```
Shouldn't we check if image is texture based before?

nikolaszimmermann replied:
Done elsewhere.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -1010,35 +1000,114 @@ void GraphicsContextSkia::drawSkiaPattern(const sk_sp<SkImage>& image, const Int
+    auto* grContext = PlatformDisplay::sharedDisplay().skiaGrContext();
+    RELEASE_ASSERT(grContext);
+
+    grContext->flush(surface);
```
And here. Also the code is the same, I'm not an expert on c++ templates, but I think we could use just one implementation using a template, no?

nikolaszimmermann replied:
The SkSurface variant is not used in this patch, but in the one converting ImageBitmap. Need to think about it.

nikolaszimmermann replied:
I will postpone the introduction of the SkSurface variant for a follow-up PR.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -1010,35 +1000,114 @@ void GraphicsContextSkia::drawSkiaPattern(const sk_sp<SkImage>& image, const Int
+void GraphicsContextSkia::createAcceleratedRenderingFenceIfNeeded(const sk_sp<SkImage>& image)
+{
+    if (!image || !image->isTextureBacked() || m_contextType != ContextType::RecordingContext)
+        return;
```
Ah, I see we check it here. Why do we need both functions?

nikolaszimmermann replied:
you mean the SkSurface and SkImage variants?

nikolaszimmermann replied:
Ah you mean via createAcceleratedRenderingFence and createAcceleratedRenderingFenceIfNeeded -- yeah it's confusing, I'll get rid of the distinction.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.h:
```diff
@@ -27,25 +27,36 @@
 class WEBCORE_EXPORT GraphicsContextSkia final : public GraphicsContext {
 public:
-    GraphicsContextSkia(SkCanvas&, RenderingMode, RenderingPurpose, CompletionHandler<void()>&& = nullptr);
+    enum class ContextType {
```
```suggestion
    enum class ContextType : bool {
```

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.h:
```diff
@@ -27,25 +27,36 @@
-    GraphicsContextSkia(SkCanvas&, RenderingMode, RenderingPurpose, CompletionHandler<void()>&& = nullptr);
+    enum class ContextType {
+        RegularContext,
+        RecordingContext
```
Maybe we could add a RenderingPurpose so that we don't need to add another parameter

nikolaszimmermann replied:
No longer a parameter, just a private `enum class ContextMode { PaintingMode, RecordingMode }`. beginRecording() switches to RecordingMode, endRecording() back to PaintingMode. No more consturctor parameter, good suggestion!

carlosgcampos commented on Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.cpp:
```diff
@@ -294,27 +256,34 @@ Ref<CoordinatedTileBuffer> SkiaPaintingEngine::paintLayer(const GraphicsLayer& l
+    auto* recordingCanvas = m_pictureRecorder.beginRecording(dirtyRect.width(), dirtyRect.height());
+    GraphicsContextSkia recordingContext(*recordingCanvas, renderingMode, RenderingPurpose::LayerBacking, nullptr, GraphicsContextSkia::ContextType::RecordingContext);
+    paintIntoGraphicsContext(layer, recordingContext, dirtyRect, contentsOpaque, contentsScale);
+    auto picture = m_pictureRecorder.finishRecordingAsPicture();
```
I have another idea. What if instead of adding a context type construction parameter and stealing the fences layer, we don't add beginRecording/endRecording? In the beginRecording we set an internal flag in the context to indicate we are recording, and endRecording can return the fences.

nikolaszimmermann replied:
That is perfect, I did that.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.cpp:
```diff
@@ -294,27 +256,34 @@ Ref<CoordinatedTileBuffer> SkiaPaintingEngine::paintLayer(const GraphicsLayer& l
 
-    auto buffer = createBuffer(renderingMode, dirtyRect.size(), contentsOpaque);
-    buffer->beginPainting();
+    auto replayCanvas = SkiaReplayCanvas::create(dirtyRect.size(), recordingContext.releaseImageFences());
```
I think it would be better to create the replay canvas in the worker thread, since it's only needed for replay, and only when rendering mode is accelerated, for cpu rendering we don't want it

nikolaszimmermann replied:
Ack. Already changed it, only think that needs to move cross-threads is the new `ImageFenceMap` (which is just a `using SkiaImageFenceMap = HashMap<const SkImage, std::unique_ptr<GLFence>>`.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.cpp:
```diff
@@ -294,27 +256,34 @@ Ref<CoordinatedTileBuffer> SkiaPaintingEngine::paintLayer(const GraphicsLayer& l
 
     auto& workerPool = renderingMode == RenderingMode::Accelerated ? *m_gpuWorkerPool.get() : *m_cpuWorkerPool.get();
-    workerPool.postTask([buffer = Ref { buffer }, displayList = WTFMove(displayList), dirtyRect]() mutable {
+    workerPool.postTask([buffer = Ref { buffer }, picture = sk_sp<SkPicture> { picture }, dirtyRect, replayCanvas = WTFMove(replayCanvas)]() mutable {
```
So we pass the fences to the lambda instead of the canvas

carlosgcampos commented on Source/WebCore/platform/graphics/skia/SkiaReplayCanvas.cpp:
```diff
@@ -0,0 +1,301 @@
+
+SkiaReplayCanvas::~SkiaReplayCanvas()
+{
+}
```
```suggestion
SkiaReplayCanvas::~SkiaReplayCanvas() = default;
```

carlosgcampos commented on Source/WebCore/platform/graphics/skia/SkiaReplayCanvas.cpp:
```diff
@@ -0,0 +1,301 @@
+    if (image->isValid(grContext))
+        return nullptr;
+
+    // TODO: Add error reporting mechanism, a failure from GetBackendTextureFromImage() should be visible / reported.
```
```suggestion
    // FIXME: Add error reporting mechanism, a failure from GetBackendTextureFromImage() should be visible / reported.
```

carlosgcampos commented on Source/WebCore/platform/graphics/skia/SkiaReplayCanvas.cpp:
```diff
@@ -0,0 +1,301 @@
+    if (!image) {
+        drawFunction(image);
+        return;
+    }
```
If waitForRenderingCompletionAndRewrapImageIfNeeded() does the null check we don't need this block

nikolaszimmermann replied:
Good idea, fixed.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/SkiaReplayCanvas.cpp:
```diff
@@ -0,0 +1,301 @@
+    if (!image) {
+        drawFunction(paint);
+        return;
+    }
```
Same here

carlosgcampos reviewed (commented):
Great job, this looks really promising.

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -1010,35 +999,97 @@ void GraphicsContextSkia::drawSkiaPattern(const sk_sp<SkImage>& image, const Int
+
+inline std::unique_ptr<GLFence> createAcceleratedRenderingFence(const sk_sp<SkImage>& image)
+{
+    if (!image)
```
This should check for `image->isTextureBacked()` too -- so we can omit the check from `createAcceleratedRenderingFenceIfNeeeded`.

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -1010,35 +999,97 @@ void GraphicsContextSkia::drawSkiaPattern(const sk_sp<SkImage>& image, const Int
+
+void GraphicsContextSkia::createAcceleratedRenderingFenceIfNeeded(const sk_sp<SkImage>& image)
+{
+    if (!image || !image->isTextureBacked() || m_contextMode != ContextMode::RecordingMode)
```
The first two conditions should be handled by createAcceleratedRenderingFence.

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -1010,35 +999,97 @@ void GraphicsContextSkia::drawSkiaPattern(const sk_sp<SkImage>& image, const Int
+        return;
+
+    auto* image = shader->isAImage(nullptr, nullptr);
+    if (!image || !image->isTextureBacked())
```
Ditto.

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/SkiaReplayCanvas.cpp:
```diff
@@ -0,0 +1,289 @@
+void SkiaReplayCanvas::invokeDrawFunctionWithPaint(const SkPaint& paint, Function<void(const SkPaint&)>&& drawFunction)
+{
+    auto* shader = paint.getShader();
+    if (!shader) {
```
We can also save this branch, if we use: `auto *image = shader ? shader->isAImage(&localMatrix, mode) : nullptr` below.

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/SkiaReplayCanvas.cpp:
```diff
@@ -0,0 +1,289 @@
+
+void SkiaReplayCanvas::invokeDrawFunctionWithShader(const SkShader* shader, Function<void(const SkShader*)>&& drawFunction)
+{
+    if (!shader) {
```
Ditto.

nikolaszimmermann reviewed (commented):
Some notes to myself:

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -1010,35 +999,91 @@ void GraphicsContextSkia::drawSkiaPattern(const sk_sp<SkImage>& image, const Int
+    return nullptr;
+}
+
+void GraphicsContextSkia::trackAcceleratedRenderingFenceIfNeeded(const sk_sp<SkImage>& image)
```
I think we can make this const if we mark the fence map as mutable, that way we don't need to remove the const from the setup functions

nikolaszimmermann replied:
Well, in the end `setupFillPaint()` needs to alter `m_imageFenceMap` -- marking it as `const` just hides the fact that as side-effect it will mutate member variables. Therefore I think omitting the const is correct - I'd usually use `mutable` as last resort, not on purpose.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.cpp:
```diff
@@ -294,27 +256,35 @@ Ref<CoordinatedTileBuffer> SkiaPaintingEngine::paintLayer(const GraphicsLayer& l
     buffer->beginPainting();
 
+    WTFBeginSignpost(this, RecordTile);
+    auto* recordingCanvas = m_pictureRecorder.beginRecording(dirtyRect.width(), dirtyRect.height());
```
Why is the SkPictureRecorder a member? We only need it here, and after finishRecordingAsPicture it can be destroyed, no?

nikolaszimmermann replied:
Agreed. I initially thought it's too expensive to re-create it every time -- especially since it has built-in logic to reuse the fRecorder it allocated in its constructor, across recording operations (see reset()).

We can avoid the member to start with, and profile it once again and eventually optimize this, if it proofs to be expensive.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.cpp:
```diff
@@ -294,27 +256,35 @@ Ref<CoordinatedTileBuffer> SkiaPaintingEngine::paintLayer(const GraphicsLayer& l
+    workerPool.postTask([buffer = Ref { buffer }, picture = sk_sp<SkPicture> { picture }, dirtyRect, imageToFenceMap = WTFMove(imageToFenceMap)]() mutable {
         if (auto* canvas = buffer->canvas()) {
+            auto replayCanvas = SkiaReplayCanvas::create(dirtyRect.size(), WTFMove(imageToFenceMap));
+            replayCanvas->addCanvas(canvas);
```
I think we should only use the SkiaReplayCanvas when doing GPU rendering. We could even avoid it if the image to fence map is empty, not sure it's worth it, though.

nikolaszimmermann replied:
Good moment. EWS is green as it-is, so we can now optimize and avoid the SkiaReplayCanvas creation altogether whenever imageFencesMap is empty -- in both CPU and GPU drawing scenarios, e.g. a basic fillRect operation won't need a SkiaReplayCanvas in GPU rendering mode either, so I think it's the cleanest approach to decide based on the image to fence map.

nikolaszimmermann committed:
[Gtk][WPE][Skia] Use SkPicture(Recorder) instead of DisplayList(Recorder) for paint command recording in multi-threaded rendering mode
https://bugs.webkit.org/show_bug.cgi?id=289456

Reviewed by Carlos Garcia Campos.

For multi-threaded rendering Gtk/WPE ports serialized the painting
commands for a given layer and tile into a DisplayList, and dispatched
a task to a WorkerPool to replay the DisplayList in a secondary thread.

The cross-thread usage of DisplayLists is fragile, and the way we use
DisplayList wouldn't work on Apple ports, due to timing differences when
a new frame rendering starts. While DisplayList (and its ResourceHeap) protects
the referenced WebCore objects (such as Gradient / Pattern etc.) from
destruction, we're not protected from mutations between the recording of
a DisplayList and the replay, later on in time.

Furthermore, the Gtk/WPE ports currently depend on specific
synchronization primitives (creating fences in the accelerated rendering
case during recording, and waiting for the fences in the replay, before
attempting to use the fence-protected GPU resource) that were built only
into DisplayList. That was a bad idea, the Skia ports should be able to
properly paint in both CPU and GPU mode, independent of whether
DisplayList is used or not.

This PR moves the Gtk/WPE ports away from DisplayList recording to use a
Skia built-in functionality instead: SkPicture and SkPictureRecorder.
While this doesn't yet solve the mutation problem in a general way, it
removes a lot of potential problems, when we start to paint the next
frame earlier, then we do right now. Furthermore it resolves some
general design issues Kimmo detected with the current code -- namely
the sprinkling of USE(SKIA) throughout ImageBuffer/ImageBufferBackend
and NativeImage & co.

Follow-up patches will remove the Skia-specific functionality that was
added to ImageBuffer & co to make threaded DisplayList rendering
possible for our ports (albeit only under limited constraints, that we
want to get rid of in the long-term).

Covered by existing tests.

* LayoutTests/fast/backgrounds/background-clip-border-area-alpha.html:
Adjust for slight pixel test differences -- no visible change.
* LayoutTests/fast/multicol/transform-inside-opacity.html:
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/root-element-background-image-transparency-001.html:
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/root-element-background-image-transparency-002.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/clip-002.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/clip-004.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/clip-005.html:
* LayoutTests/svg/as-background-image/svg-transformed-background.html:
* LayoutTests/svg/repaint/buffered-rendering-static-image.html:
* Source/WebCore/platform/Skia.cmake:
* Source/WebCore/platform/SourcesSkia.txt:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::surface const):
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
(WebCore::ImageBufferBackend::surface const):
(WebCore::ImageBufferBackend::waitForAcceleratedRenderingFenceCompletion):
* Source/WebCore/platform/graphics/NativeImage.h:
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::drawNativeImageInternal):
(WebCore::GraphicsContextSkia::setupFillSource):
(WebCore::GraphicsContextSkia::setupStrokeSource):
(WebCore::GraphicsContextSkia::clipToImageBuffer):
(WebCore::createAcceleratedSurface):
(WebCore::GraphicsContextSkia::drawPattern):
(WebCore::GraphicsContextSkia::beginRecording):
(WebCore::GraphicsContextSkia::endRecording):
(WebCore::GraphicsContextSkia::createAcceleratedRenderingFenceIfNeeded):
(WebCore::GraphicsContextSkia::trackAcceleratedRenderingFenceIfNeeded):
(WebCore::GraphicsContextSkia::drawSkiaImage): Deleted.
(WebCore::GraphicsContextSkia::drawFilteredImageBuffer): Deleted.
(WebCore::GraphicsContextSkia::setupFillSource const): Deleted.
(WebCore::GraphicsContextSkia::setupStrokeSource const): Deleted.
(WebCore::GraphicsContextSkia::drawSkiaPattern): Deleted.
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.h:
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.h:
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaSurfaceBackend.h:
* Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.cpp:
(WebCore::SkiaPaintingEngine::postPaintingTask):
(WebCore::SkiaPaintingEngine::recordDisplayList const): Deleted.
(WebCore::SkiaPaintingEngine::paintDisplayListIntoBuffer): Deleted.
* Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.h:
* Source/WebCore/platform/graphics/skia/SkiaReplayCanvas.cpp: Added.
(WebCore::SkiaReplayCanvas::SkiaReplayCanvas):
(WebCore::SkiaReplayCanvas::create):
(WebCore::SkiaReplayCanvas::waitForRenderingCompletionAndRewrapImageIfNeeded):
(WebCore::SkiaReplayCanvas::invokeDrawFunctionWithImage):
(WebCore::SkiaReplayCanvas::invokeDrawFunctionWithPaint):
(WebCore::SkiaReplayCanvas::invokeDrawFunctionWithShader):
(WebCore::SkiaReplayCanvas::onClipShader):
(WebCore::SkiaReplayCanvas::onDrawArc):
(WebCore::SkiaReplayCanvas::onDrawAtlas2):
(WebCore::SkiaReplayCanvas::onDrawBehind):
(WebCore::SkiaReplayCanvas::onDrawDRRect):
(WebCore::SkiaReplayCanvas::onDrawEdgeAAImageSet2):
(WebCore::SkiaReplayCanvas::onDrawGlyphRunList):
(WebCore::SkiaReplayCanvas::onDrawImage2):
(WebCore::SkiaReplayCanvas::onDrawImageLattice2):
(WebCore::SkiaReplayCanvas::onDrawImageRect2):
(WebCore::SkiaReplayCanvas::onDrawOval):
(WebCore::SkiaReplayCanvas::onDrawPaint):
(WebCore::SkiaReplayCanvas::onDrawPoints):
(WebCore::SkiaReplayCanvas::onDrawPatch):
(WebCore::SkiaReplayCanvas::onDrawPath):
(WebCore::SkiaReplayCanvas::onDrawRRect):
(WebCore::SkiaReplayCanvas::onDrawRect):
(WebCore::SkiaReplayCanvas::onDrawRegion):
(WebCore::SkiaReplayCanvas::onDrawSlug):
(WebCore::SkiaReplayCanvas::onDrawTextBlob):
(WebCore::SkiaReplayCanvas::onDrawVerticesObject):
* Source/WebCore/platform/graphics/skia/SkiaReplayCanvas.h: Added.

Canonical link: https://commits.webkit.org/292639@main
